### PR TITLE
Itemlink directive

### PR DESCRIPTION
**Pull request by bavovanachte**

### CODE REVIEW

File: mlx/traceability.py [hunk 1]
```diff
@@ -249,6 +253,38 @@ def run(self):
 
         return [item_list_node]
 
+class ItemLinkDirective(Directive):
+    """
+    Directive to generate a list of items.
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: copy/paste leftover

File: mlx/traceability.py [hunk 2]
```diff
@@ -718,6 +754,32 @@ def process_item_nodes(app, doctree, fromdocname):
 
         node.replace_self(new_node)
 
+    # We process the item-link items. These are stored in a dictionary, so they
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: no need to store in seperate dict here. Just parse the object, your `node` variable and use `env.traceability_collection` to add the relations using `add_relation`. Example in line 171 of this file.

File: mlx/traceability.py [hunk 3]
```diff
@@ -718,6 +754,32 @@ def process_item_nodes(app, doctree, fromdocname):
...
+    #   'SWRQT-sit':   [
+    #                   ('ITEST-amet', 'depends_on'),
+    #                   ('ITEST-ipsum', 'depends_on'),
+    #                  ]
+    #
+    # }
+    itemlinks = {}
+    for node in doctree.traversingrse(ItemLink):
+        for source in node['sources']:
+            if itemlinks[source] is None:
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: something like this:

```
for source in node['sources']:
    for target in node['target']:
        env.traceability_collection.add_relation(source, node['type'], target)
```

File: mlx/traceability.py [hunk 4]
```diff
@@ -718,6 +754,32 @@ def process_item_nodes(app, doctree, fromdocname):
...
+    #                   ('ITEST-dolor', 'verifies'),
+    #                  ]
+    #   'SWRQT-sit':   [
+    #                   ('ITEST-amet', 'depends_on'),
+    #                   ('ITEST-ipsum', 'depends_on'),
+    #                  ]
+    #
+    # }
+    itemlinks = {}
+    for node in doctree.traversingrse(ItemLink):
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: this whole block needs to be moved up in the function, we first need to add all relations, before we render the item-list/matrix/tree/item...

File: mlx/traceability.py [hunk 5]
```diff
@@ -249,6 +253,38 @@ def run(self):
...
+    # Content disallowed
+    has_content = False
+
+    def run(self):
+        env = self.state.document.settings.env
+        app = env.app
+
+        item_link_node = ItemLink('')
+
+        item_link_node['sources'] = self.options['sources']
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: `sources` is a string, you need to `split()` on spaces to get the different items (see line 477 of same file for example)

same for `targets`

File: mlx/traceability.py [hunk 6]
```diff
@@ -76,6 +76,10 @@ class ItemTree(nodes.General, nodes.Element):
     '''Tree-view on documentation items'''
     pass
 
+class ItemLink(nodes.General, nodes.Element):
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: not related to this line, but more conceptual:

note to self: are we sure all items exist (source and target) when the item-link starts adding relations to the database...

File: mlx/traceability.py [hunk 7]
```diff
@@ -249,6 +253,38 @@ def run(self):
...
+
+    def run(self):
+        env = self.state.document.settings.env
+        app = env.app
+
+        item_link_node = ItemLink('')
+
+        item_link_node['sources'] = self.options['sources']
+        item_link_node['targets'] = self.options['targets']
+        item_link_node['type'] = self.options['type']
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: if one of these arguemtns to item-link is not given, the tool will crash on the access of the options['...'].

2 solutions:
- can we make the arguments mandatory: sphinx checks and reports if the argument is missing (preferred)
- we probably can't live without these arguments, so use `report_warning` in order to inform the user

File: mlx/traceability.py (hunk 4 above)
Comment (SteinHeselmans) [reply mlx/traceability.py]: right before the self_test as we want these extra relations to be checked as well...

File: mlx/traceability.py (hunk 6 above)
Comment (bavovanachte) [reply mlx/traceability.py]: The ``add_relation`` function performs checks on the source and relationship type. For some reason it doesn't do it for the target. 

I assume this is intentional. Any idea why?

**Comment by bavovanachte** [comment]
Ready for review. CI probably still failing due to the setuptools_scm issue, but I hope that gets fixed upstream soon.

File: mlx/traceability.py (hunk 1 above)
Comment (bavovanachte) [reply mlx/traceability.py]: Fixed.

File: mlx/traceability.py (hunk 6 above)
Comment (SteinHeselmans) [reply mlx/traceability.py]: When items are defined in RST, we parse them in the plugin one by one. While parsing we update the relations of this item. So we store where the items link to = forward relation.  But we also need to store the reverse relation. So to the target item(s) we add the 'reverse' relation to the source item.

Depending on the order of parsing the RST files and items, a target item might not exist in the 'database' yet. So the script will add a placeholder item and add the 'reverse' relation to it. Later, when parsing more items the placeholder item is updated with the actual content.

A self-check performs check on the database to make sure there is no placeholder items left.

For the above reason we cannot assert that the target item is in the database so we don't check it. The source item and relation kind should exist, that is what we can assert.

This information is - or should be - in the readme somehow. At least a mention of placeholder items...

File: README.rst [hunk 8]
```diff
@@ -428,6 +428,25 @@ The relations to other documentation items can be specified as:
...
+================================================
+
+In some cases, it's useful to add relations outside of the definition of the items
+involved. In that case, you can use the ``item-link`` directive as follows
+
+.. code-block:: rest
+
+    .. item-link::
+        :sources: RQT1 RQT2
+        :targets: RQT3 RQT4 RQT5
```
Comment (SteinHeselmans) [inline README.rst]: for easier reading, let's rename the targets to TST name items?

File: mlx/traceability.py [hunk 9]
```diff
@@ -76,6 +76,10 @@ class ItemTree(nodes.General, nodes.Element):
     '''Tree-view on documentation items'''
     pass
 
+class ItemLink(nodes.General, nodes.Element):
+    '''List of documentation items'''
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: copy pase mistake

File: setup.py [hunk 10]
```diff
@@ -3,7 +3,7 @@
 from setuptools import setup, find_packages
 
 project_url = 'https://github.com/melexis/sphinx-traceability-extension'
-version = '2.5.2'
+version = '2.5.3'
```
Comment (SteinHeselmans) [inline setup.py]: this new feature deserves a 2.6.0 version :wink:

File: example/index.rst [hunk 11]
```diff
@@ -278,6 +278,14 @@ Another tree that should spawn a warning as the relation in *type* does not exis
...
         :type: fulfilled_by fulfills
 
+This is a subtitle that has a ``item-link`` item under it. You shouldn't see anything in the rendering, though
+--------------------------------------------------------------------------------------------------------------
+
+.. item-link::
+    :sources: r001
+    :targets: r002
+    :type: trace
+
```
Comment (SteinHeselmans) [inline example/index.rst]: add an example with some missing argument (sources, targets, type), it should throw a warning

File: mlx/traceability.py [hunk 12]
```diff
@@ -718,6 +754,19 @@ def process_item_nodes(app, doctree, fromdocname):
 
         node.replace_self(new_node)
 
+    # Processing of the item-link items. They get added as additional relationships
```
Comment (SteinHeselmans) [inline mlx/traceability.py]: move this up in the fuction to right below the self_test(), line 583

all relations need to be known before rendering item-matrix, etc

**Comment by SteinHeselmans** [comment]
extra warning should be removed in d2bc227

**Comment by SteinHeselmans** [comment]
final change in 4bc66af (fix problem when item-link is defined in seperate RST-document)

**Comment by bavovanachte** [comment]
Tested the latest changes: they seem to work :+1:  Ok for merge?

**Pull request merged**